File: targets/go/fiber/parameters.go
// parameters: router captures with segment depth held constant, each bound as an integer and
// echoed.
//
// Fiber's own binder fills the struct: c.Bind().URI reads the `uri:` tag from the route's
// captures. The struct is both what Fiber fills and what the handler echoes, so nothing copies
// one shape into another. The bind skips validation, because the StructValidator in the config
// runs after every bind and these structs have no rules for it.
//
// Fiber tries routes in the order they were registered. :one matches the literal segment
// static as readily as a number, so the static route is registered first.
package main

import (
	"github.com/gofiber/fiber/v3"
	d "github.com/ianjohnson/requestbench/targets/go/_shared"
)

// rb:wiring parameters.*
type oneCapture struct {
	One int `uri:"one" json:"one"`
}

// rb:wiring parameters.*
type twoCaptures struct {
	One int `uri:"one" json:"one"`
	Two int `uri:"two" json:"two"`
}

// bindCaptures fills out from the route's captures, answering the failure itself if there is
// one, in the shape bindQuery answers a query Fiber could not bind.
// rb:wiring parameters.*
func bindCaptures(c fiber.Ctx, out any) bool {
	if err := c.Bind().SkipValidation(true).URI(out); err != nil {
		_ = c.Status(400).JSON(fiber.Map{"error": "invalid_path", "detail": err.Error()})
		return false
	}
	return true
}

func registerParameters(app *fiber.App) {
	app.Get("/parameters/static/segment/literal", payload("small"))

	app.Get("/parameters/:one/segment/literal", func(c fiber.Ctx) error {
		var p oneCapture
		if !bindCaptures(c, &p) {
			return nil
		}
		return c.JSON(d.WithEcho("small", p))
	})

	app.Get("/parameters/:one/with-second/:two", func(c fiber.Ctx) error {
		var p twoCaptures
		if !bindCaptures(c, &p) {
			return nil
		}
		return c.JSON(d.WithEcho("small", p))
	})
}

File: frameworks/node/hono/Implementation/routes/parameters.ts
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";

import type { Routes } from "../app.ts";

// rb:wiring parameters.*
/**
 * The captures as each route binds them. zValidator's param target hands zod the route's captures,
 * which are strings, and zod converts each to an integer, so the handler reads numbers.
 */
const one = z.object({ one: z.coerce.number().int() });

const two = z.object({ one: z.coerce.number().int(), two: z.coerce.number().int() });
// rb:end

/**
 * parameters: router captures, each echoed back as the integer its schema converts it to. The
 * static path is registered first, because it also matches the one-capture route, and Hono runs
 * the handlers that match a path in the order they were registered.
 */
const parameters: Routes = (app, p) => {
  app.get("/parameters/static/segment/literal", (c) => c.json(p.small));

  app.get("/parameters/:one/segment/literal", zValidator("param", one), (c) => c.json({ ...p.small, echo: c.req.valid("param") }));

  app.get("/parameters/:one/with-second/:two", zValidator("param", two), (c) => c.json({ ...p.small, echo: c.req.valid("param") }));
};

export default parameters;
